README.md
# Angel.co Startup Parser
* `angel.rb`: This script parses startups on angel.co by a market tag and creates a sqlite
database with that parsed data.
* `sqlite_to_csv_converter.rb`: This script creates two csv files(one for nodes, one for
edges) with parsed data by `angel.rb` for graph commons.

# Features
* Database views

# Requirements
* Ruby >= 2.2.x
* Bundler(`gem install bundler`)
* See also [Gemfile](Gemfile)

# Usage
First of all, rename the following file names:

* `config.example.yml` => `config.yml`
* `views.example.yml` => `views.yml` (*optional, if you want add a new database view(s)* )

and edit those files by yourself. After:

## Setup
```bash
$ git clone https://erenhatirnaz@bitbucket.org/erenhatirnaz/angel.co-startup-parser.git
$ cd angel.co-startup-parser
$ bundler install
```

## Creation database
Run `ruby angel.rb -m "[Market Tag]"` or just `ruby angel.rb` command, and then follow the steps.

## Convert database to csv files (for import GraphCommons)
```bash
$ ruby sqlite_to_csv_converter.rb -d [Database directory]
```

# Output database schema
![database schema](db_schema.png)

# License
Apache 2.0


CHANGELOG.md
# Change Log

## 2.1.1
### Bugfixes:
* [Issue #3](https://bitbucket.org/erenhatirnaz/angel.co-startup-parser/issues/3#comment-33719502):
  `startup_id` and `location_id` can be same sometimes.
    * Added new controls to prevent data repetition.
### Notable changes:
* Fixed a typo in error message.
* Reordered Gemfile and forgotten Gemfile.lock updated.

## 2.1.0
### Bugfixes:
* [Issue #1](https://bitbucket.org/erenhatirnaz/angel.co-startup-parser/issues/1):
  Several properties mustn't contains whitespaces on both sides.
* [Issue #3](https://bitbucket.org/erenhatirnaz/angel.co-startup-parser/issues/3):
  `startup_id` and `location_id` can be same sometimes.
### Proposals:
* [Issue #2](https://bitbucket.org/erenhatirnaz/angel.co-startup-parser/issues/2):
  If get an error during database creation, this database directory have to be
  deleted.
* [Issue #4](https://bitbucket.org/erenhatirnaz/angel.co-startup-parser/issues/4):
  Console output methods need refactor.

## 2.0.2
### Hotfixes:
* [Issue #6](https://bitbucket.org/erenhatirnaz/angel.co-startup-parser/issues/6):
Capitalization error in the library name. 

## 2.0.1
### Hotfixes:
* [Issue #5](https://bitbucket.org/erenhatirnaz/angel.co-startup-parser/issues/5):
  If the views.yml file is missing, throws an unhandled error.

## 2.0.0
### Features:
* Database integration and database views support.
    * Added option whether the sought market tag include or not.
    * Added database directory name generator method in `helper.rb`.
    * Added `create_and_open_database` method into `helper.rb` for creation 
      database file.
    * Added models to use the tables in the created database.
    * Added `sqlite_to_csv_converter` script.
* Progress Bar: Show the CSV file creation progress.
* Colorize output messages of the console: More readable console outputs.
    * Added bundler for dependency management.

## 1.0.0
* First release.


helper.rb
require 'uri'
require 'json'
require 'net/https'
require 'yaml'
require 'English'

require 'sequel'
require 'sqlite3'

def query(query_url, access_token)
  query_url += (query_url.include?('?') ? '&' : '?') + "access_token=#{access_token}"
  url = URI.parse(query_url)
  req = Net::HTTP::Get.new(url.to_s)
  http = Net::HTTP.new(url.host, url.port)

  http.use_ssl = true
  http.verify_mode = OpenSSL::SSL::VERIFY_NONE
  response = http.request(req)

  if response.code.to_i == 401
    abort '[ERR] This access token is invalid or has been revoked by the user.'.on_red
  end

  JSON.parse(response.body)
end

def get_user_choice(collection, message)
  puts message.to_s.magenta

  collection.each_with_index do |item, index|
    puts((index + 1).to_s.light_red + " )- #{item['name']}".yellow)
  end

  print '> Select an item from the list: '.light_blue

  selected_index = gets.chomp.to_i - 1
  if selected_index < 0 || collection[selected_index].nil?
    abort '[ERR] Please enter a valid item number!'.on_red
  end

  collection[selected_index]
end

def normalize_string(string)
  string.tr("/\u2019/", "/\u0027/").tr("/\u0022/", "/\u0027/")
end

def generate_database_directory_name(market_tag_name)
  market_tag_name = market_tag_name.tr(' ', '-').downcase
  database_directory_name = nil

  loop do
    database_directory_name = "#{market_tag_name}-#{rand(1000)}"
    break unless File.exist?(database_directory_name)
  end

  database_directory_name
end

def delete_directory(directory_name)
  Dir.glob("#{directory_name}/**") { |file_name| File.delete(file_name) }
  Dir.delete(directory_name)
end

def create_and_open_database(database_path)
  db_connection = Sequel.connect("sqlite://#{database_path}")

  db_connection.create_table(:locations) do
    column :id, :integer, primary_key: true
    column :name, :text, null: false
  end

  db_connection.create_table(:markets) do
    column :id, :integer, primary_key: true
    column :name, :text, null: false
  end

  db_connection.create_table(:startups) do
    column :id, :integer, primary_key: true
    column :name, :text, null: false
    column :description, :text
    column :website_url, :text
    column :logo_url, :text
    column :reference, :text
    column :quality, :integer
    column :follower_count, :integer
  end

  db_connection.create_table(:locations_startups) do
    foreign_key :startup_id, :startups, null: false
    foreign_key :location_id, :locations, null: false
    primary_key [:startup_id, :location_id]
  end

  db_connection.create_table(:markets_startups) do
    foreign_key :startup_id, :startups, null: false
    foreign_key :market_id, :markets, null: false
    primary_key [:startup_id, :market_id]
  end

  views_file = File.exist?('views.yml') ? 'views.yml' : 'views.example.yml'
  import_views_from_yml_file(db_connection, views_file)

  Dir['./models/*.rb'].each { |model| require model }

  db_connection
end

def import_views_from_yml_file(db_connection, file_name)
  views = YAML.load_file(file_name)
  views.each do |view|
    next if (view['name'].nil? || view['name'].empty?) ||
            (view['query'].nil? || view['query'].empty?)
    begin
      db_connection.create_or_replace_view(view['name'].strip, view['query'].strip)
    rescue Sequel::DatabaseError
      puts '[ERR] There is an error in this query:'.white.on_red
      puts "\tName : ".cyan + view['name']
      puts "\tQuery: ".cyan + view['query'].strip
      puts "\tError: ".cyan + $ERROR_INFO.to_s[23..-1]
    end
  end
  fail_views_count = views.length - db_connection.views.length
  puts "DB Views -> Total  :#{views.length}\n".yellow + \
       "\t    Created:#{db_connection.views.length}\n".yellow + \
       "\t    Fail   :#{fail_views_count}".yellow
end

def open_database(database_path)
  Sequel.connect("sqlite://#{database_path}")

  Dir['./models/*.rb'].each { |model| require model }
end


sqlite_to_csv_converter.rb
require 'optparse'

require 'colorize'

require './helper.rb'

options = { databaseDirectory: nil }

parser = OptionParser.new do |opts|
  banner = []
  banner[0] = "Script Name\t: SQLite to CSV converter for GraphCommans".light_green
  banner[1] = "Description\t: This script creates two csv file(one for\n".light_green\
            + "\t\t  nodes, one for edges) with parsed data\n".light_green\
            + "\t\t  by `angel.rb` for graph commons.".light_green
  banner[2] = "Developer\t: Eren Hatirnaz <erenhatirnaz@atinasoft.com>".light_green
  banner[3] = "Company\t\t: Atinasoft\t<info@atinasoft.com>".light_green
  banner[4] = '~'.cyan * 61
  banner[5] = 'Usage: ruby sqlite_to_csv_converter.rb -d DATABASE_DIRECTORY'
  opts.banner = banner.join("\n")

  opts.on('-d', '--database-directory DATABASE_DIRECTORY',
          'The directory name containing the database created by angel.rb') do |database_directory|
    options[:databaseDirectory] = database_directory
  end

  opts.on_tail('-h', '--help', 'Displays Help') do
    puts opts
    exit
  end
end

parser.parse!

if options[:databaseDirectory].nil?
  print '> Enter database directory: '.light_blue
  options[:databaseDirectory] = gets.chomp.downcase
  if options[:databaseDirectory].empty?
    abort '[ERR] Database directory cannot be empty!'.on_red
  end
end

unless File.exist?(options[:databaseDirectory])
  abort '[ERR] Database directory not found!'.on_red
end

database_directory_name = options[:databaseDirectory].sub('/', '')
database_path = "#{database_directory_name}/database.sqlite"

open_database(database_path)

nodes_file_path = "#{database_directory_name}/nodes.csv"
edges_file_path = "#{database_directory_name}/edges.csv"

if File.exist?(nodes_file_path) || File.exist?(edges_file_path)
  puts '[WARN] nodes.csv or edges.csv file is already exist!'.black.on_yellow
  print '> Do you want to overwrite these files?(y/N): '.light_blue
  overwrite = gets.chomp.downcase
  if overwrite == 'y' || overwrite == 'yes'
    File.delete(nodes_file_path)
    File.delete(edges_file_path)
  else
    exit
  end
end

nodes_file = File.new(nodes_file_path, 'w')
nodes_file.puts('Type,Name,Description,Image,Reference,Website Url,Quality,Follower Count')

Startup.all.each do |startup|
  nodes_file.puts('Startup,'\
                + "\"#{normalize_string(startup[:name].to_s)}\","\
                + "\"#{normalize_string(startup[:description].to_s)}\","\
                + "#{startup[:logo_url]},"\
                + "#{startup[:reference]},"\
                + "#{startup[:website_url]},"\
                + "#{startup[:quality]},"\
                + startup[:follower_count].to_s)
end

Market.all.each { |market| nodes_file.puts("Market,\"#{market[:name]}\"") }
Location.all.each { |location| nodes_file.puts("Location,\"#{location[:name]}\"") }

nodes_file.close

edges_file = File.new(edges_file_path, 'w')
edges_file.puts('FROM NODE TYPE,FROM NODE NAME,EDGE TYPE,TO NODE TYPE,TO NODE NAME,Weight')

Startup.all.each do |startup|
  prefix = "Startup,\"#{normalize_string(startup[:name].to_s)}\""

  startup.markets.each { |market| edges_file.puts("#{prefix},BELONGS_TO,Market,#{market[:name]}") }
  startup.locations.each { |location| edges_file.puts("#{prefix},BASED_IN,Location,#{location[:name]}") }
end

edges_file.close

statistics = []
statistics[0] = "Total line count of nodes.csv\t:#{IO.readlines(nodes_file).count}"
statistics[1] = "Total line count of edges.csv\t:#{IO.readlines(edges_file).count}"
statistics[2] = "Directory of output CSV files\t:#{database_directory_name}"

puts '-'.cyan * 55
puts statistics.join("\n").yellow
puts "\nCSV files created succesfully!".black.on_green


angel.rb
require 'yaml'
require 'optparse'

require 'colorize'
require 'ruby-progressbar'

require './helper.rb'

options = { soughtMarketTagName: nil }

parser = OptionParser.new do |opts|
  banner = []
  banner[0] = "Script Name\t: Angel.co Startup Parser".light_green
  banner[1] = "Description\t: This script parses startups on angel.co by\n".light_green \
            + "\t\t  a market tag and creates a sqlite database\n".light_green \
            + "\t\t  with that parsed data.".light_green
  banner[2] = "Developer\t: Eren Hatirnaz <erenhatirnaz@atinasoft.com>".light_green
  banner[3] = "Company\t\t: Atinasoft \t<info@atinasoft.com>".light_green
  banner[4] = '~'.cyan * 61
  banner[5] = 'Usage: ruby angel.rb -m MARKET_TAG_NAME'
  opts.banner = banner.join("\n")

  opts.on('-m', '--market-tag-name MARKET_TAG_NAME', 'Sought market tag name') do |market_tag_name|
    options[:soughtMarketTagName] = market_tag_name
  end

  opts.on_tail('-h', '--help', 'Displays Help') do
    puts opts
    exit
  end
end

parser.parse!

unless File.exist?('config.yml')
  abort "[ERR] config.yml file not found!\n".on_red\
      + "Please, rename to file name 'config.example.yml' to 'config.yml' and edit that file by yourself.".on_red
end

config = YAML.load_file('config.yml')
ACCESS_TOKEN = config['angel-co']['access-token']

if options[:soughtMarketTagName].nil?
  print '> Enter market tag: '.light_blue
  options[:soughtMarketTagName] = gets.chomp
  if options[:soughtMarketTagName].empty?
    abort '[ERR] Market tag cannot be empty!'.on_red
  end
end

market_tag_name = options[:soughtMarketTagName]

search_results = query("https://api.angel.co/1/search?query=#{market_tag_name}&type=MarketTag", ACCESS_TOKEN)

if search_results.empty?
  abort '[WARN] No result for this market tag='.black.on_yellow + market_tag_name.black.on_white
end

selected_market_tag = search_results.count > 1 ? get_user_choice(search_results, 'Founded markets:') : search_results[0]

print '> Do you want include this market tag into database?(y/N): '.light_blue
include_main_market_tag = gets.chomp.downcase
include_main_market_tag = include_main_market_tag == 'yes' || include_main_market_tag == 'y' ? true : false

market_tag_id = selected_market_tag['id']
market_tag_name = selected_market_tag['name'].strip

database_directory_name = generate_database_directory_name(market_tag_name)
database_path = "#{database_directory_name}/database.sqlite"
Dir.mkdir(database_directory_name)

DB = create_and_open_database(database_path)

market_startups = query("https://api.angel.co/1/tags/#{market_tag_id}/startups", ACCESS_TOKEN)

last_page = market_startups['last_page']
hidden_startup_count = 0

progressbar = ProgressBar.create(format: '%a <%B> %p%% %t', total: last_page)

begin
  last_page.times do |current_page|
    market_startups = query("https://api.angel.co/1/tags/#{market_tag_id}/startups?page=#{current_page + 1}",
                            ACCESS_TOKEN)

    market_startups['startups'].each do |item|
      if item['hidden']
        hidden_startup_count += 1
        next
      end

      next unless Startup.where(name: item['name'].strip).first.nil?

      startup_description = item['high_concept'].strip if item['high_concept']

      startup = Startup.create(id: item['id'],
                               name: item['name'].strip,
                               description: startup_description,
                               website_url: item['company_url'],
                               logo_url: item['logo_url'],
                               reference: item['angellist_url'],
                               quality: item['quality'],
                               follower_count: item['follower_count'])

      item['markets'].each do |mrkt|
        next if mrkt['id'] == market_tag_id && !include_main_market_tag
        market_name = mrkt['display_name'].strip
        market = Market.where(name: market_name).first || Market.create(id: mrkt['id'], name: market_name)
        startup.add_market(market) unless startup.markets.include?(market)
      end

      item['locations'].each do |lctn|
        location_name = lctn['display_name']
        location = Location.where(name: location_name).first || Location.create(id: lctn['id'], name: location_name)
        startup.add_location(location) unless startup.locations.include?(location)
      end
    end

    progressbar.increment
  end
rescue => exception
  DB.disconnect
  delete_directory(database_directory_name)
  abort "[ERR] An error occurred: #{exception}".on_red
end

statistics = []
statistics[0] = "Total startups\t\t\t:#{Startup.all.length} (#{hidden_startup_count} hidden startup)"
statistics[1] = "Total markets\t\t\t:#{Market.all.length}"
statistics[2] = "Total locations\t\t\t:#{Location.all.length}"
statistics[3] = "Directory of output database\t:#{database_directory_name}"

puts '-'.cyan * 55
puts statistics.join("\n").yellow
puts "\nDatabase created succesfully!".black.on_green
puts "If you want CSV files, run '".black.on_green \
    + "ruby sqlite_to_csv_converter.rb -d #{database_directory_name}".red.on_white \
    + "' command.".black.on_green
